File: domain/src/main/scala/net/kemitix/thorp/domain/MD5Hash.scala
package net.kemitix.thorp.domain

import java.util.Base64

final case class MD5Hash(in: String)

object MD5Hash {
  def fromDigest(digest: Array[Byte]): MD5Hash =
    MD5Hash((digest map ("%02x" format _)).mkString)
  def hash(md5Hash: MD5Hash): String = QuoteStripper.stripQuotes(md5Hash.in)
  def digest(md5Hash: MD5Hash): Array[Byte] =
    HexEncoder.decode(MD5Hash.hash(md5Hash))
  def hash64(md5Hash: MD5Hash): String =
    Base64.getEncoder.encodeToString(MD5Hash.digest(md5Hash))
}

File: domain/src/main/java/net/kemitix/thorp/domain/MD5Hash.java
package net.kemitix.thorp.domain;

import net.kemitix.mon.TypeAlias;

import java.util.Arrays;
import java.util.Base64;
import java.util.List;
import java.util.stream.Collectors;

public class MD5Hash extends TypeAlias<List<String>> {
    private MD5Hash(List<String> value) {
        super(value);
    }
    public static MD5Hash create(String in) {
        return new MD5Hash(Arrays.asList(in.split("")));
    }
    public static MD5Hash fromDigest(Byte[] digest) {
        return new MD5Hash(
                Arrays.stream(digest)
                        .map(b -> String.format("%02x", b))
                        .collect(Collectors.toList()));
    }
    public String hash() {
        return QuoteStripper.stripQuotes(String.join("", getValue()));
    }
    public byte[] digest() {
        return HexEncoder.decode(hash());
    }
    public String hash64() {
        return Base64.getEncoder().encodeToString(digest());
    }
}
